File: sdk/java/src/main/java/com/zorroa/zmlp/sdk/domain/asset/Asset.java
package com.zorroa.zmlp.sdk.domain.asset;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.zorroa.zmlp.sdk.Json;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.function.Predicate;
import java.util.regex.Pattern;
import java.util.stream.Collectors;


/**
 * The file information and all the metadata generated during Analysis.
 */
public class Asset {

    private String id;
    private Map<String, Object> document;

    private static final Logger logger = LoggerFactory.getLogger(Asset.class);

    private static final Pattern ATTR_SPLIT_PATTERN = Pattern.compile(".", Pattern.LITERAL);

    public Asset() {
    }

    public Asset(String id) {
        this.id = id;
        document = new HashMap();
    }

    public Asset(String id, Map<String, Object> document) {
        this.id = id;
        this.document = document;
    }

    /**
     * Return all stored files associated with this asset.  Optionally
     * filter the results.
     *
     * @param name      The associated files name.
     * @param category  The associated files category, eg proxy, backup, etc.
     * @param mimetype  The mimetype must start with this string.
     * @param extension The file name must have the given extension.
     * @param attrs     The file must have all of the given attributes.
     * @return List of Dict Pixml file records.
     */

    public List getFiles(List<String> name, List<String> category, List<String> mimetype, List<String> extension, Map attrs, List attrKeys) {

        // Get Files Object
        List<Map<String, Object>> files = (List) document.getOrDefault("files", new ArrayList());

        // Create Name Filter
        Predicate<Map<String, Object>> namePredicate = f -> {
            String fileNameAttr = (String) f.get("name");
            return fileNameAttr == null ? false : name.contains(fileNameAttr);
        };

        //Create Category Filter
        Predicate<Map<String, Object>> categoryPredicate = f -> {
            String categoryAttr = (String) f.get("category");
            return category.contains(categoryAttr);
        };

        //Create mimetype Filter
        Predicate<Map<String, Object>> mimeTypePredicate = f ->
                (mimetype.parallelStream().filter((String mimeType) -> {
                    String mimetypeAttr = (String) f.get("mimetype");
                    return mimetypeAttr == null ? false : mimetypeAttr.startsWith(mimeType);
                }).collect(Collectors.toList()).size() > 0);

        //Create Extension Filter
        Predicate<Map<String, Object>> extensionPredicate = f ->
                (extension.parallelStream().filter((String ext) -> {
                    String nameAttr = (String) f.get("name");
                    return nameAttr == null ? false : nameAttr.endsWith(ext);
                })).collect(Collectors.toList()).size() > 0;

        //Create Attrs Filter
        Predicate<Map<String, Object>> attrsPredicate = f ->
                (Boolean) attrs.entrySet().stream()
                        .map((entry) -> {
                                    Map.Entry key = (Map.Entry) entry;
                                    Map attrsObject = (Map) f.get("attrs");
                                    if (attrsObject == null)
                                        return false;

                                    Object o = attrsObject.get(((Map.Entry) entry).getKey());
                                    return o == null ? false : o.equals(key.getValue());
                                }
                        ).reduce((o1, o2) -> ((Boolean) o1) && ((Boolean) o2)).orElse(false);

        // Create Attrs Keys Filter
        Predicate<Map<String, Object>> attrsKeysPredicate = f -> {
            Map attributes = (Map) f.get("attrs");
            return attributes == null ? false : attributes.keySet().containsAll(attrKeys);
        };

        // Check which of predicates will be used
        List<Predicate> elegiblePredicates = new ArrayList();
        Optional.ofNullable(name).ifPresent((ignore) -> elegiblePredicates.add(namePredicate));
        Optional.ofNullable(category).ifPresent((ignore) -> elegiblePredicates.add(categoryPredicate));
        Optional.ofNullable(mimetype).ifPresent((ignore) -> elegiblePredicates.add(mimeTypePredicate));
        Optional.ofNullable(extension).ifPresent((ignore) -> elegiblePredicates.add(extensionPredicate));
        Optional.ofNullable(attrs).ifPresent((ignore) -> elegiblePredicates.add(attrsPredicate));
        Optional.ofNullable(attrKeys).ifPresent((ignore) -> elegiblePredicates.add(attrsKeysPredicate));

        //Join All predicates
        Predicate compositePredicate = elegiblePredicates.stream().reduce(w -> true, Predicate::and);

        return (List) files.parallelStream().filter(compositePredicate).collect(Collectors.toList());

    }

    /**
     * Return all stored files associated with this asset filtered by name.
     *
     * @param name The associated files name.
     * @return List of Dict Pixml file records.
     */
    public List getFilesByName(String... name) {
        if (name == null)
            return new ArrayList();
        return this.getFiles(Arrays.asList(name), null, null, null, null, null);
    }

    /**
     * Return all stored files associated with this asset filtered by category.
     *
     * @param category The associated files category, eg proxy, backup, etc.
     * @return List of Dict Pixml file records.
     */
    public List getFilesByCategory(String... category) {
        if (category == null)
            return new ArrayList();
        return this.getFiles(null, Arrays.asList(category), null, null, null, null);
    }

    /**
     * Return all stored files associated with this asset filtered by mimetype.
     *
     * @param mimetype The mimetype must start with this string.
     * @return List of Dict Pixml file records.
     */
    public List getFilesByMimetype(String... mimetype) {
        if (mimetype == null)
            return new ArrayList();
        return this.getFiles(null, null, Arrays.asList(mimetype), null, null, null);
    }

    /**
     * Return all stored files associated with this asset filtered by extension.
     *
     * @param extension The file name must have the given extension.
     * @return List of Dict Pixml file records.
     */
    public List getFilesByExtension(String... extension) {
        if (extension == null)
            return new ArrayList();
        return this.getFiles(null, null, null, Arrays.asList(extension), null, null);
    }

    /**
     * Return all stored files associated with this asset filtered by File Attrs.
     *
     * @param attrs The file must have all of the given attributes.
     * @return List of Dict Pixml file records.
     */
    public List getFilesByAttrs(Map attrs) {
        if (attrs == null)
            return new ArrayList();
        return this.getFiles(null, null, null, null, attrs, null);
    }

    /**
     * Return all stored files associated with this asset filtered by by File Attrs Keys.
     *
     * @param attrsKey The file must have all of the given attributes.
     * @return List of Dict Pixml file records.
     */
    public List getFilesByAttrsKey(String... attrsKey) {
        if (attrsKey == null)
            return new ArrayList();
        return this.getFiles(null, null, null, null, null, Arrays.asList(attrsKey));
    }

    /**
     * Get an attribute value  by its fully qualified name.
     *
     * @param attr
     * @param <T>
     * @return
     */
    public <T> T getAttr(String attr) {
        Object current = getContainer(attr, false);
        return (T) getChild(current, attrName(attr));
    }

    /**
     * Get an attribute value by its fully qualified name.  Uses a JSON mapper
     * to map the data into the specified class.
     *
     * @param attr
     * @param type
     * @param <T>
     * @return
     */
    public <T> T getAttr(String attr, Class<T> type) {
        Object current = getContainer(attr, false);
        return Json.mapper.convertValue(getChild(current, attrName(attr)), type);
    }

    /**
     * Get an attribute value by its fully qualified name.  Uses a JSON mapper
     * to map the data into the specified TypeReference.
     *
     * @param attr
     * @param type
     * @param <T>
     * @return
     */
    public <T> T getAttr(String attr, TypeReference<T> type) {
        Object current = getContainer(attr, false);
        return Json.mapper.convertValue(getChild(current, attrName(attr)), type);
    }

    /**
     * Return true if the document has the given namespace.
     *
     * @param attr
     * @return
     */
    public boolean attrExists(String attr) {
        Object container = getContainer(attr, false);
        return getChild(container, attrName(attr)) != null;
    }

    public void setAttr(String attr, Object value) {
        Object current = getContainer(attr, true);
        try {
            String key = attr.substring(attr.lastIndexOf('.') + 1);
            Map<String, Object> map = (Map) current;
            if (value == null) {
                map.put(key, null);
            } else {
                map.put(key, Json.mapper.readValue(Json.asJson(value), Object.class));
            }
        } catch (ClassCastException e) {
            throw new IllegalArgumentException("Invalid attribute, " + attr + " is not a map.");
        } catch (StringIndexOutOfBoundsException e) {
            throw new IllegalArgumentException("Invalid attribute, " + attr, e);
        } catch (JsonMappingException e) {
            throw new IllegalArgumentException("Invalid attribute, " + attr + ", cannot be serialized", e);
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException("Invalid attribute, " + attr + ", cannot be serialized", e);
        }
    }


    public boolean removeAttr(String attr) {
        Object current = getContainer(attr, true);
        try {
            String key = attr.substring(attr.lastIndexOf('.') + 1);
            Map<String, Object> map = (Map) current;
            return map.remove(key) != null;
        } catch (ClassCastException e) {
            throw new IllegalArgumentException("Invalid attribute, " + attr + " is not a map.");
        } catch (StringIndexOutOfBoundsException e) {
            throw new IllegalArgumentException("Invalid attribute, " + attr, e);
        }

    }

    /**
     * @param attr
     * @param forceExapand
     * @return
     */
    private Object getContainer(String attr, boolean forceExapand) {
        String[] parts = ATTR_SPLIT_PATTERN.split(attr);
        Object current = document;
        for (int i = 0; i < parts.length - 1; i++) {
            Object child = getChild(current, parts[i]);
            if (child == null) {
                if (forceExapand) {
                    child = createChild(current, parts[i]);
                } else {
                    return null;
                }
            }
            current = child;
        }
        return current;
    }

    /**
     * @param container
     * @param key
     * @return
     */
    private Object getChild(Object container, String key) {
        if (container == null) {
            return null;
        }
        try {
            Map<String, Object> map = (Map<String, Object>) container;
            return map.get(key);
        } catch (ClassCastException ex) {
            return null;
        }
    }

    private Object createChild(Object parent, String key) {
        Map<String, Object> result = new HashMap<>();
        try {
            Map<String, Object> map = (Map<String, Object>) parent;
            map.put(key, result);
        } catch (ClassCastException ex) {
            throw new IllegalArgumentException("Invalid attribute, parent of " + key + " is not a map.");
        }
        return result;
    }

    private String attrName(String attr) {
        return attr.substring(attr.lastIndexOf('.') + 1);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Asset)) return false;
        Asset asset = (Asset) o;
        return id.equals(asset.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    public String getId() {
        return id;
    }

    public Map<String, Object> getDocument() {
        return document;
    }
}

File: services/archivist/archivist/src/main/kotlin/com/zorroa/archivist/domain/Asset.kt
package com.zorroa.archivist.domain

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.module.kotlin.convertValue
import com.google.common.hash.Hashing
import com.zorroa.archivist.security.getProjectId
import com.zorroa.archivist.util.randomString
import com.zorroa.zmlp.util.Json
import io.swagger.annotations.ApiModel
import io.swagger.annotations.ApiModelProperty
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer
import java.security.MessageDigest
import java.util.Base64
import java.util.UUID
import java.util.regex.Pattern

@ApiModel("AssetState",
    description = "Describes the different states can asset can be in.")
enum class AssetState {

    @ApiModelProperty("The Asset been created in the database but is pending analysis.")
    Pending,

    @ApiModelProperty("The Asset has been analyzed and augmented with fields.")
    Analyzed
}

@ApiModel("Asset Spec",
    description = "Defines all the properties required to create an Asset.")
class AssetSpec(

    @ApiModelProperty("The URI location of the asset.")
    var uri: String,

    @ApiModelProperty("Additional metadata fields to add to the Asset in key/value format.")
    var attrs: Map<String, Any>? = null,

    @ApiModelProperty("Optional clip metadata specifies the portion of the asset to process.")
    var clip: Clip? = null,

    @ApiModelProperty("Optional DataSet label which puts the asset in the given DataSet.")
    val label: DataSetLabel? = null,

    @ApiModelProperty("An optional unique ID for the asset to override the auto-generated ID.")
    val id: String? = null,

    @ApiModelProperty("A hidden option for sending the entire composed document.  ", hidden = true)
    val document: MutableMap<String, Any>? = null
)

@ApiModel("Asset Counters", description = "Stores the types of asset counters we keep track off.")
class AssetCounters(

    @ApiModelProperty("Total number of assets.")
    val total: Int = 0,

    @ApiModelProperty("Total number of assets created.")
    val created: Int = 0,

    @ApiModelProperty("Total number of assets replaced")
    val replaced: Int = 0
)

@ApiModel("Asset",
    description = "The file information and all the metadata generated during Analysis.")
open class Asset(

    @ApiModelProperty("The unique ID of the Asset.")
    val id: String,

    @ApiModelProperty("The assets metadata.")
    val document: MutableMap<String, Any>
) {

    constructor() : this(randomString(24), mutableMapOf())

    constructor(id: String) : this(id, mutableMapOf())

    constructor(document: MutableMap<String, Any>) :
        this(randomString(24), document)

    /**
     * Remove an attribute.  If the attr cannot be remove it is set to null.
     *
     * @param attr
     */
    fun removeAttr(attr: String): Boolean {
        val current = getContainer(attr, true)
        val key = Attr.name(attr)

        /*
         * Finally, just try treating it like a map.
         */
        try {
            return (current as MutableMap<String, Any>).remove(key) != null
        } catch (ex: ClassCastException) {
            throw IllegalArgumentException("Invalid attribute: $attr")
        }
    }

    /**
     * Get an attribute value  by its fully qualified name.
     *
     * @param attr
     * @param <T>
     * @return
    </T> */
    fun <T> getAttr(attr: String): T? {
        val current = getContainer(attr, false)
        return getChild(current, Attr.name(attr)) as T?
    }

    /**
     * Return true if the attribute is not null.
     *
     * @param attr The attr name
     * @return true if attr exists
     */
    fun attrExists(attr: String): Boolean {
        val current = getContainer(attr, false)
        return getChild(current, Attr.name(attr)) != null
    }

    /**
     * Get an attribute value by its fully qualified name.  Uses a JSON mapper
     * to map the data into the specified class.
     *
     * @param attr
     * @param type
     * @param <T>
     * @return
    </T> */
    fun <T> getAttr(attr: String, type: Class<T>): T? {
        val current = getContainer(attr, false)
        return Json.Mapper.convertValue(getChild(current, Attr.name(attr)), type)
    }

    /**
     * Get an attribute value by its fully qualified name.  Uses a JSON mapper
     * to map the data into the specified TypeReference.
     *
     * @param attr
     * @param type
     * @param <T>
     * @return
    </T> */
    fun <T> getAttr(attr: String, type: TypeReference<T>): T? {
        val current = getContainer(attr, false)
        return Json.Mapper.convertValue(getChild(current, Attr.name(attr)), type)
    }

    /**
     * Set an attribute value.
     *
     * @param attr
     * @param value
     */
    fun setAttr(attr: String, value: Any?) {
        val current = getContainer(attr, true)
        val key = Attr.name(attr)

        try {
            if (value == null) {
                (current as MutableMap<String, Any?>)[key] = null
            } else {
                (current as MutableMap<String, Any?>)[key] = Json.Mapper.convertValue(value)
            }
        } catch (ex: ClassCastException) {
            throw IllegalArgumentException("Invalid attribute: $attr", ex)
        }
    }

    /**
     * Return true if the asset has been analayzed before.
     */
    fun isAnalyzed(): Boolean {
        return attrExists("system.timeCreated") && getAttr<String>("system.state") == AssetState.Analyzed.name
    }

    private fun getContainer(attr: String, forceExpand: Boolean): Any? {
        val parts = PATTERN_ATTR.split(attr)

        var current: Any? = document
        for (i in 0 until parts.size - 1) {

            var child = getChild(current, parts[i])
            if (child == null) {
                if (forceExpand) {
                    child = createChild(current, parts[i])
                } else {
                    return null
                }
            }
            current = child
        }
        return current
    }

    private fun getChild(`object`: Any?, key: String): Any? {
        if (`object` == null) {
            return null
        }
        return try {
            (`object` as Map<String, Any>)[key]
        } catch (ex: ClassCastException) {
            null
        }
    }

    private fun createChild(parent: Any?, key: String): Any {
        val result = mutableMapOf<String, Any>()
        try {
            (parent as MutableMap<String, Any>)[key] = result
        } catch (ex: ClassCastException) {
            throw IllegalArgumentException("Invalid attribute: $key parent: $parent")
        }
        return result
    }

    override fun toString(): String {
        return "<Asset $id - $document>"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Asset) return false

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

    companion object {

        private val logger = LoggerFactory.getLogger(Asset::class.java)

        private val PATTERN_ATTR = Pattern.compile(Attr.DELIMITER, Pattern.LITERAL)
    }
}

object Attr {

    const val DELIMITER = "."

    /**
     * Return the last part of an attribute string.  For example, if fully qualified
     * name is "a:b:c:d", this method will return "d".
     *
     * @param attr
     * @return
     */
    fun name(attr: String): String {
        return attr.substring(attr.lastIndexOf(DELIMITER) + 1)
    }
}

/**
 * A utility class for building unique Asset ids.
 *
 * @property spec: The AssetSpec
 */
class AssetIdBuilder(val spec: AssetSpec) {

    val projectId = getProjectId()
    var length = 32
    var checksum: Int? = null

    /**
     * Convert the givne UUID into a byte array.
     */
    private fun uuidToByteArray(uuid: UUID): ByteBuffer {
        val bb: ByteBuffer = ByteBuffer.wrap(ByteArray(16))
        bb.putLong(uuid.mostSignificantBits)
        bb.putLong(uuid.leastSignificantBits)
        return bb.asReadOnlyBuffer()
    }

    /**
     * Apply a checksum to ID generation.
     */
    fun checksum(bytes: ByteArray): AssetIdBuilder {
        val hf = Hashing.adler32()
        checksum = hf.hashBytes(bytes).asInt()
        return this
    }

    /**
     * Build and return the unique ID.
     */
    fun build(): String {
        if (spec.id != null) {
            return spec.id
        }

        /**
         * Nothing about the order of these statements
         * can ever change or duplicate assets will be
         * created.
         */
        val digester = MessageDigest.getInstance("SHA-256")
        digester.update(spec.uri.toByteArray())
        digester.update(uuidToByteArray(projectId))

        spec.clip?.let {
            digester.update(it.type.toByteArray())
            it.timeline?.let { timeline ->
                digester.update(timeline.toByteArray())
            }
            val buf = ByteBuffer.allocate(16)
            buf.putDouble(it.start)
            buf.putDouble(it.stop)
            digester.update(buf.array())
        }

        checksum?.let {
            val buf = ByteBuffer.allocate(4)
            buf.putInt(it)
            digester.update(buf.array())
        }
        // Clamp the size to 32, 48 is bit much and you still
        // get much better resolution than a UUID.  We could
        // also up it on shared indexes but probably not necessary.
        return Base64.getUrlEncoder()
            .encodeToString(digester.digest()).trim('=').substring(0, length)
    }
}
